File: cmd/2015/day10/main.py
import sys

day_input = sys.stdin.read().strip()


def rle_encode(s: str) -> str:
    res = ""
    cnt, prev = None, None
    for ch in s:
        if not cnt and not prev:
            prev = ch
            cnt = 1
        elif ch != prev:
            res += f"{cnt}{prev}"
            prev = ch
            cnt = 1
        else:
            cnt += 1
    if cnt > 0:
        res += f"{cnt}{prev}"
    return res


# Part 1
part_1 = day_input
i = 0
while i < 40:
    part_1 = rle_encode(part_1)
    i += 1

# Part 2
part_2 = day_input
i = 0
while i < 50:
    part_2 = rle_encode(part_2)
    i += 1

print("Part 1:", len(part_1))
print("Part 2:", len(part_2))

File: cmd/2015/day10/main.go
package main

import (
	"fmt"
	"os"
	"strings"

	"github.com/artsaban/advent-of-code/internal/solution"
)

func rleEncode(s string) string {
	format := "%d%c"
	builder := strings.Builder{}
	prev, prevCnt := '\n', -1

	for _, ch := range s {
		if prev == '\n' {
			prev = ch
			prevCnt = 1
		} else if prev != ch {
			builder.WriteString(fmt.Sprintf(format, prevCnt, prev))
			prev = ch
			prevCnt = 1
		} else {
			prevCnt += 1
		}
	}

	if prevCnt > 0 {
		builder.WriteString(fmt.Sprintf(format, prevCnt, prev))
	}

	return builder.String()
}

func part1(input string) int {
	res := input
	for range 40 {
		res = rleEncode(res)
	}
	return len(res)
}

func part2(input string) int {
	res := input
	for range 50 {
		res = rleEncode(res)
	}
	return len(res)
}

func main() {
	solution.NewSolution(os.Stdin, part1, part2).Print()
}
